### ISSUE TITLE
BSP Documentation

### ISSUE THREAD

**Issue by Traderain** [skyrim/hlviewer.js#6]
The bsp file format should be documented so the project can be easier understood. Something like what I did for the .dem format.

**Comment by skyrim** [comment]
It would be nice, but I don't have time right now. I'm busy working. You're free to do it if you like.
I used this[1] specification document, but I only used it to extract entities, textures and geometry.

Important things that I haven't done yet:

- Lightmaps
Lightmap lump stores static lighting textures. I haven't yet implemented algorithm to use them. That's why there are no shadows and other light effects on maps. [2]

- Visibility
Visibility lump has data that can be used to avoid rendering parts of the map that are not visible from current position of the camera. On small and medium maps, such as de_dust2, bkz_goldbhop, etc., it doesn't make much difference, but on maps like kz_world, nks_pupu, kz_gigablockier, etc., it would greatly improve performance. I get only ~15 fps on nks_pupu, but in-game it's stable 100 fps. [3]

[1] http://hlbsp.sourceforge.net/index.php?content=bspdef
[2] http://imgur.com/a/GPh8a (hlviewer left, in-game right)
[3] http://imgur.com/a/WpcsV
